sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +211,281 @@ object Hash {
+
+    protected def addWithValueBuilder[T](
+        value: T,
+        valueBuilderFn: ValueHashBuilder => T => ValueHashBuilder,
+    ): this.type = {
+      discard(valueBuilderFn(valueBuilder)(value))
+      this
+    }
+
+    protected def addGlobalKeyWithMaintainers(
+        globalKeyWithMaintainers: GlobalKeyWithMaintainers
+    ): this.type =
+      withContext("Maintainers")(_.addStringSet(globalKeyWithMaintainers.maintainers))
+        .withContext("Template Id")(
+          _.addQualifiedName(globalKeyWithMaintainers.globalKey.templateId.qualifiedName)
+        )
+        .withContext("Package Name")(_.addString(globalKeyWithMaintainers.globalKey.packageName))
+        .withContext("Key")(
+          _.addWithValueBuilder(globalKeyWithMaintainers.globalKey.key, _.addTypedValue)
+        )
+
+    def addNode(node: Node, nodes: Map[NodeId, Node]): NodeBuilder
+  }
+
+  private object NodeBuilder {
+    private[lf] val LanguageVersionToNodeBuilder
+        : Map[LanguageVersion, (MessageDigestWithContext, ValueHashBuilder) => NodeBuilder] = Map(
+      LanguageVersion.v2_1 -> ((messageDigest, valueBuilder) =>
+        new NodeBuilderV1(messageDigest, valueBuilder),
+      )
+    )
+
+    @throws[NodeHashingError]
+    private[lf] def builderForVersion(
+        version: LanguageVersion,
+        messageDigest: MessageDigestWithContext,
+        valueBuilder: ValueHashBuilder,
+    ): NodeBuilder =
+      NodeBuilder.LanguageVersionToNodeBuilder
+        .getOrElse(version, throw NodeHashingError.UnsupportedLanguageVersion(version))
+        .apply(messageDigest, valueBuilder)
+
+    private[lf] def builderForNode(
+        node: Node,
+        messageDigest: MessageDigestWithContext,
+        valueBuilder: ValueHashBuilder,
+    ): NodeBuilder = node.optVersion match {
+      case Some(version) => builderForVersion(version, messageDigest, valueBuilder)
+      // If the node is version agnostic, the builder doesn't matter so pick V1
+      // This only applies to rollback nodes
+      case None => new NodeBuilderV1(messageDigest, valueBuilder)
+    }
+  }
+
+  private final class NodeBuilderV1(
+      messageDigest: MessageDigestWithContext,
+      override protected val valueBuilder: ValueHashBuilder,
+  ) extends NodeBuilder(NodeHashVersion.V1, messageDigest, valueBuilder) {
+    private def addCreateNode(create: Node.Create): this.type = {
+      addContext("Create Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(create.coid, _.addCid))
+        .withContext("Package Name")(_.addString(create.packageName))
+        .withContext("Template Id")(_.addIdentifier(create.templateId))
+        .withContext("Arg")(_.addWithValueBuilder(create.arg, _.addTypedValue))
+        .withContext("Signatories")(_.addStringSet(create.signatories))
+        .withContext("Stakeholders")(_.addStringSet(create.stakeholders))
+        .withContext("Global Keys")(_.addOptional(create.keyOpt, _.addGlobalKeyWithMaintainers))
```

Comment:
Do we want to add the `version` field of nodes here? We already add the "hashing version" here https://github.com/digital-asset/daml/pull/20172/files#diff-6cdb90c1758150a46910f9e1a187569615c4f998358046ad930f317bde3f50e5R282-R285
But they might be different. Typically the hashing version could stay the same across different LF versions if nothing warrants a change in the hashing algorithm.